### ISSUE TITLE
Serious regression with latest commits

### ISSUE THREAD

**Issue by bpecsek** [marcoheisig/sb-simd#15]
Hi Marco,

I am seeing serious regression with the latest commits.
The speed of the [spectral-norm #2 ](https://github.com/hanabi1224/Programming-Language-Benchmarks/blob/main/bench/algorithm/spectral-norm/2.cl) code is almost 4 x slower now.

Old:
CL-USER> (time (main 11000))
1.274224153
Evaluation took:
  0.735 seconds of real time
  2.876537 seconds of total run time (2.852310 user, 0.024227 system)
  391.43% CPU
  2,064,257,985 processor cycles
  462,160 bytes consed

New:
CL-USER> (time (main 11000))
1.274224153
Evaluation took:
  2.728 seconds of real time
  10.785834 seconds of total run time (10.785834 user, 0.000000 system)
  395.38% CPU
  7,667,022,426 processor cycles
  428,816 bytes consed

You can see the immense speed drop on the [Benchmark Site](https://programming-language-benchmarks.vercel.app/problem/spectral-norm) as well for 2.cl and 7.cl.

**Comment by bpecsek** [comment]
I've just checked and commit 8506eae93c2e53f8bac3c0dd5a895b5b396583fc is causing the slowdown.

**Comment by bpecsek** [comment]
Please see the generated assy for eval-a-times-u

Old (fast):
 ```
CL-USER> (disassemble 'eval-a-times-u)
; disassembly for EVAL-A-TIMES-U
; Size: 217 bytes. Origin: #x53CDAC82                         ; EVAL-A-TIMES-U
; C82:       C5FB106A01       VMOVSD XMM5, [RDX+1]
; C87:       E9BB000000       JMP L3
; C8C:       0F1F4000         NOP
; C90: L0:   498BC0           MOV RAX, R8
; C93:       48D1F8           SAR RAX, 1
; C96:       C5F957C0         VXORPD XMM0, XMM0, XMM0
; C9A:       C4E1FB2AC0       VCVTSI2SD XMM0, XMM0, RAX
; C9F:       C4E27D19C8       VBROADCASTSD YMM1, XMM0
; CA4:       0F280595FFFFFF   MOVAPS XMM0, [RIP-107]          ; [#x53CDAC40]
; CAB:       0F28159EFFFFFF   MOVAPS XMM2, [RIP-98]           ; [#x53CDAC50]
; CB2:       C4E37D18C201     VINSERTF128 YMM0, YMM0, YMM2, 1
; CB8:       C5F558C0         VADDPD YMM0, YMM1, YMM0
; CBC:       C5FC28E0         VMOVAPS YMM4, YMM0
; CC0:       C5FC100D38FFFFFF VMOVUPS YMM1, [RIP-200]         ; [#x53CDAC00]
; CC8:       C5FD58C9         VADDPD YMM1, YMM0, YMM1
; CCC:       C5EC57D2         VXORPS YMM2, YMM2, YMM2
; CD0:       C5FD58D2         VADDPD YMM2, YMM0, YMM2
; CD4:       C5FC57C0         VXORPS YMM0, YMM0, YMM0
; CD8:       C5F558C0         VADDPD YMM0, YMM1, YMM0
; CDC:       C5FC101D3CFFFFFF VMOVUPS YMM3, [RIP-196]         ; [#x53CDAC20]
; CE4:       C5FD59C3         VMULPD YMM0, YMM0, YMM3
; CE8:       C5ED59C0         VMULPD YMM0, YMM2, YMM0
; CEC:       C5FD58C9         VADDPD YMM1, YMM0, YMM1
; CF0:       C4E27D19C5       VBROADCASTSD YMM0, XMM5
; CF5:       C5FD5EC1         VDIVPD YMM0, YMM0, YMM1
; CF9:       B902000000       MOV ECX, 2
; CFE:       EB37             JMP L2
; D00: L1:   C5FB105C8A01     VMOVSD XMM3, [RDX+RCX*4+1]
; D06:       488BC1           MOV RAX, RCX
; D09:       48D1F8           SAR RAX, 1
; D0C:       C5E957D2         VXORPD XMM2, XMM2, XMM2
; D10:       C4E1EB2AD0       VCVTSI2SD XMM2, XMM2, RAX
; D15:       C4E27D19D2       VBROADCASTSD YMM2, XMM2
; D1A:       C5DD58D2         VADDPD YMM2, YMM4, YMM2
; D1E:       C5F558D2         VADDPD YMM2, YMM1, YMM2
; D22:       C5FC28CA         VMOVAPS YMM1, YMM2
; D26:       C4E27D19DB       VBROADCASTSD YMM3, XMM3
; D2B:       C5E55ED2         VDIVPD YMM2, YMM3, YMM2
; D2F:       C5FD58C2         VADDPD YMM0, YMM0, YMM2
; D33:       4883C102         ADD RCX, 2
; D37: L2:   4839D9           CMP RCX, RBX
; D3A:       7CC4             JL L1
; D3C:       C4A17D11448701   VMOVUPD [RDI+R8*4+1], YMM0
; D43:       4983C008         ADD R8, 8
; D47: L3:   4939F0           CMP R8, RSI
; D4A:       0F8C40FFFFFF     JL L0
; D50:       BA17011050       MOV EDX, #x50100117             ; NIL
; D55:       488BE5           MOV RSP, RBP
; D58:       F8               CLC
; D59:       5D               POP RBP
; D5A:       C3               RET
```

New (slow):
```
CL-USER> (disassemble 'eval-a-times-u)
; disassembly for EVAL-A-TIMES-U
; Size: 217 bytes. Origin: #x53AEF202                         ; EVAL-A-TIMES-U
; 02:       C5FB106A01       VMOVSD XMM5, [RDX+1]
; 07:       E9BB000000       JMP L3
; 0C:       0F1F4000         NOP
; 10: L0:   498BC0           MOV RAX, R8
; 13:       48D1F8           SAR RAX, 1
; 16:       660F57C0         XORPD XMM0, XMM0    <<<------ used to be VEX
; 1A:       F2480F2AC0       CVTSI2SD XMM0, RAX  <<<------ used to be VEX
; 1F:       C4E27D19C8       VBROADCASTSD YMM1, XMM0
; 24:       0F280595FFFFFF   MOVAPS XMM0, [RIP-107]           ; [#x53AEF1C0]
; 2B:       0F28159EFFFFFF   MOVAPS XMM2, [RIP-98]            ; [#x53AEF1D0]
; 32:       C4E37D18C201     VINSERTF128 YMM0, YMM0, YMM2, 1
; 38:       C5F558C0         VADDPD YMM0, YMM1, YMM0
; 3C:       C5FC28E0         VMOVAPS YMM4, YMM0
; 40:       C5FC100D38FFFFFF VMOVUPS YMM1, [RIP-200]          ; [#x53AEF180]
; 48:       C5FD58C9         VADDPD YMM1, YMM0, YMM1
; 4C:       C5EC57D2         VXORPS YMM2, YMM2, YMM2
; 50:       C5FD58D2         VADDPD YMM2, YMM0, YMM2
; 54:       C5FC57C0         VXORPS YMM0, YMM0, YMM0
; 58:       C5F558C0         VADDPD YMM0, YMM1, YMM0
; 5C:       C5FC101D3CFFFFFF VMOVUPS YMM3, [RIP-196]          ; [#x53AEF1A0]
; 64:       C5FD59C3         VMULPD YMM0, YMM0, YMM3
; 68:       C5ED59C0         VMULPD YMM0, YMM2, YMM0
; 6C:       C5FD58C9         VADDPD YMM1, YMM0, YMM1
; 70:       C4E27D19C5       VBROADCASTSD YMM0, XMM5
; 75:       C5FD5EC1         VDIVPD YMM0, YMM0, YMM1
; 79:       B902000000       MOV ECX, 2
; 7E:       EB37             JMP L2
; 80: L1:   C5FB105C8A01     VMOVSD XMM3, [RDX+RCX*4+1]
; 86:       488BC1           MOV RAX, RCX
; 89:       48D1F8           SAR RAX, 1
; 8C:       660F57D2         XORPD XMM2, XMM2    <<<------ used to be VEX
; 90:       F2480F2AD0       CVTSI2SD XMM2, RAX  <<<------ used to be VEX
; 95:       C4E27D19D2       VBROADCASTSD YMM2, XMM2
; 9A:       C5DD58D2         VADDPD YMM2, YMM4, YMM2
; 9E:       C5F558D2         VADDPD YMM2, YMM1, YMM2
; A2:       C5FC28CA         VMOVAPS YMM1, YMM2
; A6:       C4E27D19DB       VBROADCASTSD YMM3, XMM3
; AB:       C5E55ED2         VDIVPD YMM2, YMM3, YMM2
; AF:       C5FD58C2         VADDPD YMM0, YMM0, YMM2
; B3:       4883C102         ADD RCX, 2
; B7: L2:   4839D9           CMP RCX, RBX
; BA:       7CC4             JL L1
; BC:       C4A17D11448701   VMOVUPD [RDI+R8*4+1], YMM0
; C3:       4983C008         ADD R8, 8
; C7: L3:   4939F0           CMP R8, RSI
; CA:       0F8C40FFFFFF     JL L0
; D0:       BA17011050       MOV EDX, #x50100117              ; NIL
; D5:       488BE5           MOV RSP, RBP
; D8:       F8               CLC
; D9:       5D               POP RBP
; DA:       C3               RET
```

**Comment by marcoheisig** [comment]
Hi Bela,

thanks for notifying me.  That looks like a problem with an easy fix, I'll work on it.

That commit is actually pretty cool, because it cuts down the compilation time of sb-simd and all the code using it by a lot.  But it seems the instruction set information gets lost somewhere in the process.

Best regards,
Marco

**Comment by bpecsek** [comment]
I've noticed the huge drop in compilation time and got very happy until I realized the slow down.
There must be an easy fix for it retaining the good compilation speed with the correct instruction set information.

**Comment by marcoheisig** [comment]
The regression should be fixed in 966014ad8dec2b5157082c3471801a9689c69efe.

**Comment by bpecsek** [comment]
Speed is back and compilation speed is still fast!!!
Perfect.
